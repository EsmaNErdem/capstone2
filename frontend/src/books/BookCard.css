.BookCard {
    width: 100%;
    max-width: 200px;
    margin: 10px;
    padding: 10px;
    border: 1px solid #ddd;
    border-radius: 8px;
    overflow: hidden;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
    transition: transform 0.3s ease-in-out;
  
    &:hover {
      transform: scale(1.02);
    }
  }
  
  .CardImage img {
    width: 100%;
    height: 150px; 
    object-fit: cover;
  }
  
  .CardContent {
    padding: 16px;
  }

  .Title {
    text-decoration: none;
  }

  h2 {
    color: #a074f3; 
    font-size: 1.2rem;
    margin-bottom: 8px;
  }

  h2:hover {
    color: #ff6347; 
    transition: color 0.3s ease;
  }
  
  h3 {
    font-size: .85rem;
    color: gray;
    margin-bottom: 12px;
    font-style: italic;
  }
  
  p {
    font-size: 1rem;
    margin-bottom: 12px;
  }

  .CardFooter {
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin-top: 10px; 
  }

  .CardImage {
    border-radius: 8px;
  }
  
  @media (min-width: 768px) {
    .BookCard {
      max-width: 100%;
    }
  
    .CardImage img {
      height: 200px;
    }
  }
  
  @media (min-width: 992px) {
    .BookCard {
      max-width: calc(20%); /* Four books in a row with margin */
    }
  }
  
