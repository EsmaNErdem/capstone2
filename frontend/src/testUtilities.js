import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
