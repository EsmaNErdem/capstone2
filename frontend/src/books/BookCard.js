import React, { useState } from "react";
import { Link } from 'react-router-dom';
import useBookLike from "../hooks/useBookLike";
import useReviewAdd from "../hooks/useReviewAdd";
import useReviewDelete from "../hooks/useReviewDelete";
import Alert from "../utilities/Alert"
import BookClubApi from "../api";
import BookReviewDrawer from "./BookReviewDrawer";
import IconButton from '@mui/material/IconButton';
import FavoriteIcon from '@mui/icons-material/Favorite';
import CommentIcon from '@mui/icons-material/Comment';
import './BookCard.css'; 

/**
 * Show short info about a book
 * 
 * - Rendered by BookList to display card for each book and shows how many reviews and likes the book has.
 * - Send user like on a book. If fails, shows error to user
 * - Display book review count and like count
 * 
 * - BookList ==> BookCard
 */
const BookCard = ({ id, title, author, description, publisher, category, cover, bookLikeCount=0, reviews=[], likedBook=false, setBooks=null, currUser=false }) => {
    console.debug("BookCard");
        
    const [bookReviews, setBookReviews] = useState(reviews);
    const [isReviewDrawerOpen, setReviewDrawerOpen] = useState(false);
    const [error, setError] = useState(null);

    const { liked, likes, error: likeError, handleLikeBook } = useBookLike(id, bookLikeCount, { id, title, author, description, publisher, category, cover });
    const { error: addError, addBookReview } = useReviewAdd(setBookReviews, { id, title, author, description, publisher, category, cover }, setReviewDrawerOpen)
    const { error: deleteError, deleteBookReview } = useReviewDelete(setBookReviews);

    // open side drawer to show book related reviews
    const openReviewDrawer = async () => {
        try {
            const reviews = await BookClubApi.getAllReviewsByBook(id)
            setBookReviews(reviews)
            setReviewDrawerOpen(true);
        } catch (e) {
            console.error("BookList API call data loading error:", e);
            setError("An error occurred while fetching book reviews.");
        }
    }

    // closes drawer with book related reviews
    const closeReviewDrawer = () => {
        setReviewDrawerOpen(false);
    }

    // sends book likes to database. if BookCard component is called from Profile component, it updates user liked-book list.
    const handleLikeBookButton = () => {
        handleLikeBook();
        if(liked && currUser) {
            setBooks(prevBooks => {
              const updatedBooks = prevBooks.filter(book => book.book_id !== id)
              return updatedBooks
            })
        }
    }

    return (
        <div className="BookCard">
            <Link to={`/books/${id}`} className="Title">
                <img alt={title} src={cover} className="CardImage"/>
            </Link>
            <div className="CardContent">
                <Link to={`/books/${id}`} className="Title" data-testid="book-title-link">
                    <h2 data-testid="book-title">{title}</h2>
                </Link>
                <h3 data-testid="book-author">by {author}</h3>
                <p>{description?.length > 250 ? `${description.slice(0, 250)}...` : description}</p>
                <div className="CardFooter">
                        {!likedBook &&
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <span data-testid="review-count">{bookReviews.length}</span>
                            <IconButton  data-testid="review-button" onClick={openReviewDrawer} style={{ color: bookReviews.length > 0 ? "yellowgreen" : ""}}>
                                <CommentIcon  />
                            </IconButton>
                        </div>}
                        <div style={{ display: 'flex', alignItems: 'center' }}>
                            <span data-testid="book-card-like-count">{likes}</span>
                            <IconButton onClick={handleLikeBookButton} color={liked ? 'error' : 'default'}  data-testid="book-card-like-button">
                                <FavoriteIcon/>
                            </IconButton>
                        </div>
                </div>
                {(likeError || addError || deleteError) && <Alert type="danger" messages={[likeError || addError || deleteError]} />}
                {error ? <Alert type="danger" messages={[error]} />: null}
            </div>
            <BookReviewDrawer isOpen={isReviewDrawerOpen} onClose={closeReviewDrawer} reviews={bookReviews} addReviews={addBookReview} deleteReview={deleteBookReview} bookData ={{id, title, author, cover}} />
        </div>
    )
}

export default BookCard;