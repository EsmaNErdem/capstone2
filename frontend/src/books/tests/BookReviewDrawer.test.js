import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, act } from '@testing-library/react';
import BookCard from '../BookCard';

const mockReviews = [
  {
    reviewId: '1',
    review: 'Great book!',
    username: 'JohnDoe',
    userImg: 'user1.jpg',
    date: '2023-05-16',
    reviewLikeCount: 10,
  },
];

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        getAllReviewsByBook: jest.fn(),
    },
}));
  
afterEach(() => {
jest.clearAllMocks();
});


const book = {
    id: '1',
    title: 'Book1',
    author: 'Author1',
    publisher: 'Publisher1',
    description: 'Description1',
    category: 'Category1',
    cover: 'Cover1',
    bookLikeCount: 5,
    reviews: ["Review1"]
}

test('render BookReviewDrawer and shows book detail', async () => {
    await act(async () => {
        ({ getByTestId } = render(
            <MemoryRouter>
                <UserProvider>
                    <BookCard
                          id={book.id}
                          title={book.title}
                          author={book.author}
                          description={book.description}
                          publisher={book.publisher}
                          category={book.category}
                          cover={book.cover}
                          bookLikeCount={book.bookLikeCount}
                          reviews={book.reviews} />
                </UserProvider>
            </MemoryRouter>
        ));
    });
    
    expect(getByTestId("drawer-book-title")).toHaveTextContent("Book1")
    const bookLink = getByTestId('drawer-book-link');
    expect(bookLink.getAttribute('href')).toBe('/books/1');
});

