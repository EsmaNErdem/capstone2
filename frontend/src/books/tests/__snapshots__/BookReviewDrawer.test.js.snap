// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookReviewDrawer adds new review 1`] = `<DocumentFragment />`;

exports[`BookReviewDrawer it renders and matches with snapshot 1`] = `<DocumentFragment />`;
